### ISSUE TITLE
Wireguard support

### ISSUE THREAD

**Issue by coldreaverz** [jaagr/polybar#1531]
I saw a bridge patch was supplied and I would like to see the same functionality for wireguard interfaces
I'm not really a C programmer and tried hacking around with tuntap and the bridge patch, however it seems wireguard interfaces do not support ethtool.

```
ethtool -i wg0
Cannot get driver information: Operation not supported

```

There are different ways to detect it such as /sys/net however it seems much of polybar relies on ethtool, it would need a rewrite of quite a bit of code i suppose. Doing so would however add support for many more interfaces including linkspeed for bridge interfaces. Using the physical devices for linkspeed is not a solution for me because I have the physical devices in a different namespace which cant be read without root.

I'll be glad to help but not sure what would be the best way.

Any ideas?

**Comment by patrick96** [comment]
Since wireguard doesn't support `ethtool` it wouldn't nicely fit in with our dichotomy of wired and wireless interfaces.

We could do some decoupling, similar to what was described in #1529, and have a network adapter for each type of interface. And also define some common functionality all interfaces have (e.g. data read from `/sys/class/net`)

For future reference, so that we know what kind of data the wireguard interfaces provide, could you post the entire output of the following command, when you have a connected wireguard interface.
```bash
find -L /sys/class/net/wg0 -maxdepth 2 -type f -print0 2>/dev/null | xargs -0 -I '{}' sh -c 'echo "{}: $(cat {})"'
```

The linux ABI documentation may also be useful here: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/ABI/testing/sysfs-class-net

If you're up for it, you could come up with a good abstraction for how the network module could support different types of interfaces without having too much interface-specific code inside the module. Some interface specific code is unavoidable like handling the `%essid%` token for wireless interfaces. This would involve figuring out how we could adapt the `network` class in `net.hpp` to act as a common supertype for the adapters for all the types of network interfaces we want to support.

**Comment by Aerion** [comment]
Not op, but here is the output of the command when connected to a wireguard interface.

```
/sys/class/net/wg0/uevent: DEVTYPE=wireguard
INTERFACE=wg0
IFINDEX=4
/sys/class/net/wg0/carrier_changes: 0
/sys/class/net/wg0/carrier: 1
/sys/class/net/wg0/dev_id: 0x0
/sys/class/net/wg0/carrier_down_count: 0
/sys/class/net/wg0/proto_down: 0
/sys/class/net/wg0/address: 
/sys/class/net/wg0/operstate: unknown
/sys/class/net/wg0/phys_switch_id: 
/sys/class/net/wg0/link_mode: 0
/sys/class/net/wg0/dormant: 0
/sys/class/net/wg0/statistics/tx_errors: 0
/sys/class/net/wg0/statistics/rx_length_errors: 0
/sys/class/net/wg0/statistics/rx_packets: 480195
/sys/class/net/wg0/statistics/tx_carrier_errors: 0
/sys/class/net/wg0/statistics/tx_dropped: 989
/sys/class/net/wg0/statistics/rx_missed_errors: 0
/sys/class/net/wg0/statistics/rx_over_errors: 0
/sys/class/net/wg0/statistics/tx_aborted_errors: 0
/sys/class/net/wg0/statistics/rx_crc_errors: 0
/sys/class/net/wg0/statistics/rx_frame_errors: 0
/sys/class/net/wg0/statistics/rx_nohandler: 0
/sys/class/net/wg0/statistics/tx_fifo_errors: 0
/sys/class/net/wg0/statistics/multicast: 0
/sys/class/net/wg0/statistics/tx_packets: 342076
/sys/class/net/wg0/statistics/tx_window_errors: 0
/sys/class/net/wg0/statistics/rx_bytes: 625262576
/sys/class/net/wg0/statistics/collisions: 0
/sys/class/net/wg0/statistics/rx_dropped: 0
/sys/class/net/wg0/statistics/tx_bytes: 42285528
/sys/class/net/wg0/statistics/tx_heartbeat_errors: 0
/sys/class/net/wg0/statistics/rx_fifo_errors: 0
/sys/class/net/wg0/statistics/rx_errors: 0
/sys/class/net/wg0/statistics/tx_compressed: 0
/sys/class/net/wg0/statistics/rx_compressed: 0
/sys/class/net/wg0/mtu: 1420
/sys/class/net/wg0/gro_flush_timeout: 0
/sys/class/net/wg0/power/runtime_active_time: 0
/sys/class/net/wg0/power/runtime_active_kids: 0
/sys/class/net/wg0/power/runtime_usage: 0
/sys/class/net/wg0/power/runtime_status: unsupported
/sys/class/net/wg0/power/autosuspend_delay_ms: 
/sys/class/net/wg0/power/async: disabled
/sys/class/net/wg0/power/runtime_suspended_time: 0
/sys/class/net/wg0/power/runtime_enabled: disabled
/sys/class/net/wg0/power/control: auto
/sys/class/net/wg0/carrier_up_count: 0
/sys/class/net/wg0/ifalias: 
/sys/class/net/wg0/speed: 
/sys/class/net/wg0/netdev_group: 0
/sys/class/net/wg0/phys_port_name: 
/sys/class/net/wg0/ifindex: 4
/sys/class/net/wg0/broadcast: 
/sys/class/net/wg0/type: 65534
/sys/class/net/wg0/dev_port: 0
/sys/class/net/wg0/phys_port_id: 
/sys/class/net/wg0/name_assign_type: 3
/sys/class/net/wg0/duplex: 
/sys/class/net/wg0/addr_assign_type: 0
/sys/class/net/wg0/addr_len: 0
/sys/class/net/wg0/tx_queue_len: 1000
/sys/class/net/wg0/iflink: 4
/sys/class/net/wg0/flags: 0x91
cat: /sys/class/net/wg0/phys_switch_id: Operation not supported
cat: /sys/class/net/wg0/power/autosuspend_delay_ms: Input/output error
cat: /sys/class/net/wg0/speed: Invalid argument
cat: /sys/class/net/wg0/phys_port_name: Operation not supported
cat: /sys/class/net/wg0/phys_port_id: Operation not supported
cat: /sys/class/net/wg0/duplex: Invalid argument
```

**Comment by patrick96** [comment]
@Aerion thanks! Is this interface connected? Because if so, it will be quite difficult to add support for it, since nothing in `/sys/class/net` tells us anything about the interface.

**Comment by Aerion** [comment]
Yes, the interface was/is connected, all of my traffic was/is passing through it.

**Comment by coldreaverz** [comment]
Wireguard has no up or down state, its suppose to be a stealth vpn and sends no data unless there is something to send, it doesn't even send/receive anything when you initiate connect.

That means up/down state cant be read, but stuff like rx_bytes and tx_bytes is still usefull for upspeed and downspeed I suppose?

**Comment by patrick96** [comment]
> but stuff like rx_bytes and tx_bytes is still usefull for upspeed and downspeed I suppose?

Yeah this could probably be used.

Does it have an IP address?

**Comment by patrick96** [comment]
Hmm, thinking about it, we should able to get the ip address as we do for all interfaces with `getifaddrs()`.

**Comment by Aerion** [comment]
To confirm, yes, running the example defined in the section `Program source` of `getifaddrs(3)` displays correctly the ip address.

```
[...]
wg0      AF_INET (2)
  address: <10.0.0.2>
[...]
```

**Comment by Swivelgames** [comment]
Threw up a PR (#2980) that adds rudimentary support. Nothing crazy, since there isn't much of a way to determine the state of a wireguard interface. But it should suffice for now.

Essentially, the network module doesn't fail if it can't find the interface now. It'll simply stay in the "disconnected" state until it can find the interface. Then, it'll determine the state by probing the interface as usual. In the case of wireguard, it'll simply check for the existence of the interface and consider it connected.

The logic for each of these have been placed in their appropriate locations.

**Comment by Swivelgames** [comment]
It's been a year and some change since [the PR](https://github.com/polybar/polybar/pull/2980) was initially put out. So, to celebrate it's indefinite stagnation... :sweat_smile: 

**Arch Linux Users** can get this feature now via [`polybar-wireguard-git`](https://aur.archlinux.org/packages/polybar-wireguard-git)

I update my machine daily, and try to be timely when updating this branch as soon as I see a new version of polybar, so it should stay up to date and in sync with the upstream 👍
